parachain-tracer/src/tracker.rs

```diff
@@ -423,51 +471,48 @@ impl SubxtTracker {
 		metrics: &impl PrometheusMetrics,
 		storage: &TrackerStorage,
 	) {
-		if self.current_candidate.is_idle() {
-			progress.events.push(ParachainConsensusEvent::SkippedSlot);
-			stats.on_skipped_slot(progress);
-			metrics.on_skipped_slot(progress);
-		}
-
-		if self.current_candidate.is_backed() {
-			if let Some(candidate_hash) = self.current_candidate.candidate_hash {
-				progress.events.push(ParachainConsensusEvent::Backed(candidate_hash));
-				stats.on_backed();
-				metrics.on_backed(self.para_id);
-			}
-		}
-
-		if self.current_candidate.is_pending() || self.current_candidate.is_included() {
-			progress.bitfield_health.max_bitfield_count = self.current_candidate.max_availability_bits;
-			progress.bitfield_health.available_count = self.current_candidate.current_availability_bits;
-			progress.bitfield_health.bitfield_count = self.current_candidate.bitfield_count;
-
-			if self.current_candidate.is_data_available() {
-				if let Some(candidate_hash) = self.current_candidate.candidate_hash {
-					progress.events.push(ParachainConsensusEvent::Included(
-						candidate_hash,
-						self.current_candidate.current_availability_bits,
-						self.current_candidate.max_availability_bits,
-					));
-
-					let backed_in = self.candidate_backed_in(candidate_hash, storage).await;
-					let relay_block = self.current_relay_block.expect("Checked by caller; qed");
-					stats.on_included(relay_block.num, self.previous_included_at.map(|v| v.num), backed_in);
-					metrics.on_included(
-						relay_block.num,
-						self.previous_included_at.map(|v| v.num),
-						backed_in,
-						time_diff(Some(relay_block.ts), self.previous_included_at.map(|v| v.ts)),
-						self.para_id,
-					);
-				}
-			} else if self.is_slow_availability() {
-				progress.events.push(ParachainConsensusEvent::SlowAvailability(
-					self.current_candidate.current_availability_bits,
-					self.current_candidate.max_availability_bits,
-				));
-				stats.on_slow_availability();
-				metrics.on_slow_availability(self.para_id);
+		for candidate in self.all_candidates() {
+			match candidate.state {
+				ParachainBlockState::Idle => {
+					progress.events.push(ParachainConsensusEvent::SkippedSlot);
+					stats.on_skipped_slot(progress);
+					metrics.on_skipped_slot(progress);
+				},
+				ParachainBlockState::Backed =>
+					if let Some(candidate_hash) = candidate.candidate_hash {
```

Comment:
Could we remove the option ? Doesn't seem to make sense to have the state backed but no hash.

Reply:
Above we match `ParachainBlockState::Idle` where candidate has no hash.